data/src/main/java/com/microsoft/azure/kusto/data/CloseParentResourcesStream.java

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data;
+
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.impl.client.CloseableHttpClient;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * This class allows parent resources (HttpClient and HttpResponse) to be closed when the stream is closed.
+ * This becomes necessary when passing an InputStream externally, as postToStreamingOutput does. The parent resources
+ * must be closed, but this cannot happen until the InputStream is consumed.
+ * It takes a stream, and simply forwards the call to all of its methods, except for the {@code close()} method - which does nothing.
+ */
+public class CloseParentResourcesStream extends InputStream {
+    private final InputStream innerStream;
+    private final CloseableHttpClient httpClient;
+    private final CloseableHttpResponse httpResponse;
+
+    public CloseParentResourcesStream(InputStream innerStream, CloseableHttpClient httpClient, CloseableHttpResponse httpResponse) {
+        this.innerStream = innerStream;
+        this.httpClient = httpClient;

```

Comment:
Does this mean that this is creating a new HttpClient on every request? would it be better to re-use the same client if possible (if no concurrent requests are being made,  like HttpClientFactory in Kusto.Data)

Reply:
I agree, and even saw in the Apache HttpClient docs: "Generally it is recommended to have a single instance of HttpClient per communication component or even per application. However, if the application makes use of HttpClient only very infrequently, and keeping an idle instance of HttpClient in memory is not warranted, it is highly recommended to explicitly shut down the multithreaded connection manager prior to disposing the HttpClient instance. This will ensure proper closure of all HTTP connections in the connection pool."

However, the existing POST is implemented this way, and since this feature was requested ASAP, I considered fixing that to be out of scope.

I'll add as an issue for a future enhancement that both of these should reuse the HttpClient. It indeed supports concurrent requests, but we'll still need a Factory because it can be created with a specified timeout, and we'd need to create an HttpClient per-timeout.

WDYT?

Reply:
Per @ohadbitt 's suggestion, I'll also add a TODO to this effect.